plutus-tx-plugin/test/Strictness/Spec.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE Strict              #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# OPTIONS_GHC -fplugin PlutusTx.Plugin #-}
+
+module Strictness.Spec where
+
+import Test.Tasty.Extras
+
+import PlutusTx qualified as Tx
+import PlutusTx.Code
+import PlutusTx.Prelude qualified as PlutusTx
+import PlutusTx.Test
+import PlutusTx.TH (compile)
+
+tests :: TestNested
+tests =
+  testNested
+    "Strictness"
+    [ goldenEvalCekCatch "lambda-default" $ [lambdaDefault `unsafeApplyCode` bot]
+    , goldenPirReadable "lambda-default" lambdaDefault
+    , goldenUPlcReadable "lambda-default" lambdaDefault
+
+    -- FIXME: This should not crash, but it currently does.
+    , goldenEvalCekCatch "lambda-nonstrict" $ [lambdaNonStrict `unsafeApplyCode` bot]
+    , goldenPirReadable "lambda-nonstrict" lambdaNonStrict
+    , goldenUPlcReadable "lambda-nonstrict" lambdaNonStrict
+
+    , goldenEvalCekCatch "lambda-strict" $ [lambdaStrict `unsafeApplyCode` bot]
+    , goldenPirReadable "lambda-strict" lambdaStrict
+    , goldenUPlcReadable "lambda-strict" lambdaStrict
+
+    -- FIXME: This should crash (because the `Strict` extension is on), but it currently doesn't.
+    , goldenEvalCekCatch "let-default" $ [letDefault `unsafeApplyCode` one]
+    , goldenPirReadable "let-default" letDefault
+    , goldenUPlcReadable "let-default" letDefault
+
+    , goldenEvalCekCatch "let-nonstrict" $ [letNonStrict `unsafeApplyCode` one]
+    , goldenPirReadable "let-nonstrict" letNonStrict
+    , goldenUPlcReadable "let-nonstrict" letNonStrict
+
+    -- FIXME: This should crash, but it currently doesn't.
+    , goldenEvalCekCatch "let-strict" $ [letStrict `unsafeApplyCode` one]
+    , goldenPirReadable "let-strict" letStrict
+    , goldenUPlcReadable "let-strict" letStrict
+    ]
+
+lambdaDefault :: CompiledCode (Integer -> Integer -> Integer)
+lambdaDefault =
+  $$( compile
+        [||
+        \n -> \m -> n PlutusTx.+ m
+        ||]
+    )
+
+lambdaNonStrict :: CompiledCode (Integer -> Integer -> Integer)
+lambdaNonStrict =
+  $$( compile
+        [||
+        \(~n) -> \m -> n PlutusTx.+ m
```

Comment:
Interesting. I think we probably just can't support this. The non-strictness is gone by the time we get to Core, so then it's an issue of the fact that our function application is strict by default.

Generally, I suspect ~ patterns will work much worse than ! patterns. I think we might want to end up saying "use `Strict` and don't expect ~ patterns to work".

Reply:
> The non-strictness is gone by the time we get to Core

With the `Strict` extension, GHC Core should distinguish `n` against `~n`. GHC does perform some non-semantic-preserving optimizations, but they should be rare and can be turned off by `-fpedantic-bottoms`, so I'm optimistic about this.

Reply:
> Generally, I suspect ~ patterns will work much worse than ! patterns.

I thought `~` would work with `Strict` more or less as well as `!` without `Strict`? No idea, though.

Reply:
> I thought ~ would work with Strict more or less as well as ! without Strict?

Yes, that's what I thought.